src/tests/regression/reported/issue-422-send.f90

```diff
@@ -0,0 +1,87 @@
+! BSD 3-Clause License
+!
+! Copyright (c) 2016, Sourcery Institute
+! All rights reserved.
+!
+! Redistribution and use in source and binary forms, with or without
+! modification, are permitted provided that the following conditions are met:
+!
+! * Redistributions of source code must retain the above copyright notice, this
+!   list of conditions and the following disclaimer.
+!
+! * Redistributions in binary form must reproduce the above copyright notice,
+!   this list of conditions and the following disclaimer in the documentation
+!   and/or other materials provided with the distribution.
+!
+! * Neither the name of the copyright holder nor the names of its
+!   contributors may be used to endorse or promote products derived from
+!   this software without specific prior written permission.
+!
+! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+! AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+! IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+! DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+! FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+program main
+  !! author: Damian Rouson
+  !! date: 2017-08-01
+  !! category: regression
+  !! Test whether assigning an allocatable coarray component to an allocatable
+  !! coarray component works.
+  !! OpenCoarrays issue #422
+  use iso_fortran_env, only : error_unit
+  implicit none
+
+  type foo
+    !! A derived type is required to demonstrate issue #422
+    integer, allocatable :: bar(:)[:]
+  end type
+  type(foo) :: foobar
+
+  enum, bind(C)
+    enumerator :: recipient=1, provider
+  end enum
+
+  integer, parameter :: message(*)=[3,4]
+    !! Data for provider image to put on recipient image
+
+  associate( me=>this_image(), N_images=>num_images(), required_images=>maxval([recipient,provider]) )
+
+    verify_num_images: if (N_images<required_images) then
+      write(error_unit,*)  "issue-422-send.f90 requires at least ",required_images," images"
+      error stop
+    end if verify_num_images
+
+    allocate(foobar%bar(size(message))[*],source=me)
```

Comment:
I agree that we *shouldn't* need it, but sadly shouldn't isn't the same as don't: https://github.com/sourceryinstitute/OpenCoarrays/issues/243#issuecomment-272920512 #243 was fixed on GCC trunk, and needs to be back ported to make it into GFortran < 7. I can confirm that if you repeat the test enough times with GCC 6, it will trigger the race condition and fail.